module/apmotel/config.go

```diff
@@ -0,0 +1,87 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//go:build go1.18
+// +build go1.18
+
+package apmotel // import "go.elastic.co/apm/module/apmotel"
+
+import (
+	"log"
+
+	"go.opentelemetry.io/otel/sdk/metric"
+	"go.opentelemetry.io/otel/sdk/metric/aggregation"
+)
+
+var customHistogramBoundaries = []float64{
+	0.00390625, 0.00552427, 0.0078125, 0.0110485, 0.015625, 0.0220971, 0.03125,
+	0.0441942, 0.0625, 0.0883883, 0.125, 0.176777, 0.25, 0.353553, 0.5, 0.707107,
+	1, 1.41421, 2, 2.82843, 4, 5.65685, 8, 11.3137, 16, 22.6274, 32, 45.2548, 64,
+	90.5097, 128, 181.019, 256, 362.039, 512, 724.077, 1024, 1448.15, 2048,
+	2896.31, 4096, 5792.62, 8192, 11585.2, 16384, 23170.5, 32768, 46341.0, 65536,
+	92681.9, 131072,
+}
+
+type config struct {
+	aggregation metric.AggregationSelector
+}
+
+type Option func(config) config
+
+func newConfig(opts ...Option) config {
+	cfg := config{
+		aggregation: customAggregationSelector,
+	}
+	for _, opt := range opts {
+		opt(cfg)
+	}
+
+	return cfg
+}
+
+func (cfg config) manualReaderOptions() []metric.ManualReaderOption {
+	opts := []metric.ManualReaderOption{}
+	opts = append(opts, metric.WithAggregationSelector(cfg.aggregation))
+	return opts
+}
+
+// WithAggregationSelector configure the Aggregation Selector the exporter will
+// use. If no AggregationSelector is provided the DefaultAggregationSelector is
+// used.
+func WithAggregationSelector(agg metric.AggregationSelector) Option {
+	return func(cfg config) config {
+		cfg.aggregation = agg
+		return cfg
+	}
+}
+
+func customAggregationSelector(ik metric.InstrumentKind) aggregation.Aggregation {
+	switch ik {
+	case metric.InstrumentKindCounter, metric.InstrumentKindUpDownCounter, metric.InstrumentKindObservableCounter, metric.InstrumentKindObservableUpDownCounter:
+		return aggregation.Sum{}
+	case metric.InstrumentKindObservableGauge:
+		return aggregation.LastValue{}
+	case metric.InstrumentKindHistogram:
+		return aggregation.ExplicitBucketHistogram{
+			Boundaries: customHistogramBoundaries,
+			NoMinMax:   false,
+		}
+	}
+
+	log.Panicf("unknown instrument kind %q", ik)
```

Comment:
I'm worried about this since instrumented applications will crash from the instrumentation code. I don't think that's what we want, right? Could we maybe just log, print or ignore altogether? I'm unsure what the current recommendation would be.

Reply:
OpenTelemetry itself used a `panic` if that happens. I changed it to a `log.Panicf` so we could mention which instrument kind is at fault.

https://github.com/open-telemetry/opentelemetry-go/blob/65ebe5e50f1c4d5aeb01d9693551bc4aab888149/sdk/metric/reader.go#L147-L159

Reply:
If a new instrument kind is somehow introduced, customers could get logs, but those wouldn't be actionable to them.
Introducing new instrument kinds would also break a lot of instrumentations, so I think that would probably cause a major release of otel anyway.

Reply:
Would it make sense to return [`aggregation.Default`](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/metric/aggregation#Default) in this case? If there's no sensible option and opentelemetry-go will panic anyway, I guess it's OK, but I'd also prefer not to panic if we can avoid it.

Reply:
There's no way to test it, since we can't create an instrumentation kind which doesn't exist. But If I use it for sum and count, I'm getting errors:

```
could not create all aggregators: creating aggregator with instrumentKind: 2, aggregation {}: unrecognized aggregation: {}
```

```
func customAggregationSelector(ik metric.InstrumentKind) aggregation.Aggregation {
	switch ik {
	case metric.InstrumentKindHistogram:
		return aggregation.ExplicitBucketHistogram{
			Boundaries: customHistogramBoundaries,
			NoMinMax:   false,
		}
	default:
		return aggregation.Default{}
	}
}
```

Which may be an otel bug. The error is raised here, since `aggregation.Default{}` doesn't match anything in the case.
https://github.com/open-telemetry/opentelemetry-go/blob/65ebe5e50f1c4d5aeb01d9693551bc4aab888149/sdk/metric/pipeline.go#L391

Reply:
I have opened a PR on otel to fix this issue: https://github.com/open-telemetry/opentelemetry-go/pull/3967
And changed this method to rely otel's default behavior (which panics as well)